File: src/net/merayen/elastic/backend/architectures/local/nodes/delay_1/LProcessor.java
package net.merayen.elastic.backend.architectures.local.nodes.delay_1;

import net.merayen.elastic.backend.architectures.local.LocalProcessor;
import net.merayen.elastic.backend.architectures.local.lets.AudioInlet;
import net.merayen.elastic.backend.architectures.local.lets.AudioOutlet;
import net.merayen.elastic.util.Postmaster;

public class LProcessor extends LocalProcessor {
	private Delay[] delays;

	@Override
	protected void onInit() {}

	@Override
	protected void onPrepare() {}

	@Override
	protected void onProcess() {
		AudioInlet input = (AudioInlet)getInlet("input");
		AudioOutlet output = (AudioOutlet)getOutlet("output");

		if(output != null) {
			if(input != null) {
				int start = input.read;
				int stop = input.outlet.written;
				if(stop - start > 0) {

					int channelCount = input.outlet.getChannelCount();

					ensureDelayBuffers(channelCount);

					output.setChannelCount(channelCount);

					for (int channel = 0; channel < channelCount; channel++) {
						int position = delays[channel].process(input.outlet.audio[channel], start, stop);
						for (int i = start; i < stop; i++)
							output.audio[channel][i] = delays[channel].buffer[position++ % delays[channel].buffer.length];
					}

					output.written = stop;
					input.read = stop;
					output.push();
				}
			} else {
				output.written = buffer_size;
				output.push();
			}
		} else if(input != null) {
			input.read = input.outlet.written;
		}
	}

	@Override
	protected void onMessage(Postmaster.Message message) {}

	@Override
	protected void onDestroy() {}

	private void ensureDelayBuffers(int channelCount) {
		if(delays == null || delays.length != channelCount) {
			delays = new Delay[channelCount];
			for(int i = 0; i < channelCount; i++) {
				delays[i] = new Delay(sample_rate * 4); // Allows 1 second of delay

				// DEBUG. Take these parameters from the UI instead
				final int TAPS = 100;
				for(int lol = 0; lol < TAPS; lol++)
					delays[i].addTap(new Delay.Tap((int)(Math.floor(Math.random() * sample_rate * 2)), 0.5f, 0.99f / TAPS));

				/*delays[i].addTap(new Delay.Tap(sample_rate / 4, 0.0f, .2f));
				delays[i].addTap(new Delay.Tap(sample_rate / 2, 0.0f, .3f));*/
			}
		}
	}
}

File: src/main/kotlin/net/merayen/elastic/backend/architectures/local/nodes/histogram_1/LProcessor.kt
package net.merayen.elastic.backend.architectures.local.nodes.histogram_1

import net.merayen.elastic.backend.architectures.local.LocalProcessor
import net.merayen.elastic.backend.architectures.local.lets.AudioInlet

class LProcessor : LocalProcessor() {

	/**
	 * Width of the histogram.
	 * Set by the LNode.
	 */
	var width = 2f

	/**
	 * Where the histogram lowest value starts.
	 * Set by the LNode.
	 */
	var start = -1f

	/**
	 * How many bins.
	 * Set by the LNode.
	 */
	var resolution = 100

	/**
	 * Maximum value read in current frame
	 */
	var maxValue = 0f

	/**
	 * Minimum value read in current frame
	 */
	var minValue = 0f

	var buckets = FloatArray(0)
		private set

	override fun onInit() {}
	override fun onPrepare() {
		if (buckets.size != resolution) {
			buckets = FloatArray(resolution)
		} else {
			for (i in 0 until resolution)
				buckets[i] = 0f
		}

		minValue = Float.MAX_VALUE
		maxValue = Float.MIN_VALUE
	}

	override fun onProcess() {
		val dataPort = getInlet("data") ?: return

		val available = dataPort.available()

		if (!dataPort.available())
			return

		if (dataPort is AudioInlet) {
			for (channel in 0 until dataPort.outlet.audio.size) {
				val audio = dataPort.outlet.audio[channel] ?: continue

				for (i in 0 until buffer_size) {
					val sample = audio[i]
					if ((sample > 0.01f || sample < -0.01) && sample >= start && sample < start+width)
						buckets[((sample - start) / width * (resolution-1)).toInt()] += 1f

					if (sample < minValue)
						minValue = sample
					else if (sample > maxValue)
						maxValue = sample
				}
			}
		}
	}

	override fun onDestroy() {}
}